File: backend/app/schemas/groups.py
"""그룹/멤버 API 요청·응답 스키마 (PRD 6.4)."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import GroupRole


class GroupCreateRequest(BaseModel):
    """그룹 생성 (PRD 6.4 POST /api/groups)."""

    name: str = Field(min_length=1, max_length=100)
    description: str | None = None


class GroupUpdateRequest(BaseModel):
    """그룹 정보 수정 (PRD 6.4 PUT /api/groups/{id}). 부분 갱신."""

    name: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = None


class GroupResponse(BaseModel):
    """그룹 기본 정보."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None = None
    owner_user_id: int
    is_active: bool
    created_at: datetime
    updated_at: datetime


class GroupSummaryResponse(GroupResponse):
    """그룹 목록 항목 — 멤버 수와 요청자의 소속 역할 포함 (PRD 6.4 GET /api/groups)."""

    member_count: int
    my_role: str | None = None


class GroupListResponse(BaseModel):
    """페이지네이션된 그룹 목록."""

    items: list[GroupSummaryResponse]
    total: int
    page: int
    size: int


class GroupMemberResponse(BaseModel):
    """활성 그룹원 (이메일/표시명/역할/가입일)."""

    user_id: int
    email: str
    display_name: str
    role: str
    joined_at: datetime


class GroupMemberListResponse(BaseModel):
    """그룹원 목록 (PRD 6.4 GET /api/groups/{id}/members)."""

    group_id: int
    items: list[GroupMemberResponse]


class GroupDetailResponse(GroupSummaryResponse):
    """그룹 상세 — 기본 정보 + 활성 멤버 목록 (PRD 6.4 GET /api/groups/{id})."""

    members: list[GroupMemberResponse]


class MemberInviteRequest(BaseModel):
    """그룹원 초대 (PRD 6.4 POST /api/groups/{id}/members). role 생략 시 member."""

    user_id: int
    role: GroupRole = GroupRole.MEMBER


class MemberRoleUpdateRequest(BaseModel):
    """그룹원 역할 변경 (PRD 6.4 PUT .../members/{userId}/role). admin|member 만."""

    role: GroupRole


class TransferOwnershipRequest(BaseModel):
    """소유권 이전 (PRD 6.4 POST /api/groups/{id}/transfer-ownership)."""

    user_id: int

File: frontend/src/api/groups.ts
/** 그룹/멤버 관리 API (PRD 6.4). */

import apiClient from "./client";
import type {
  Group,
  GroupDetail,
  GroupListResponse,
  GroupMember,
  GroupRole,
} from "./types";

export async function listGroups(page = 1, size = 50): Promise<GroupListResponse> {
  const { data } = await apiClient.get<GroupListResponse>("/groups", {
    params: { page, size },
  });
  return data;
}

export async function getGroup(id: number): Promise<GroupDetail> {
  const { data } = await apiClient.get<GroupDetail>(`/groups/${id}`);
  return data;
}

export async function createGroup(name: string, description?: string): Promise<Group> {
  const { data } = await apiClient.post<Group>("/groups", {
    name,
    description: description || null,
  });
  return data;
}

export async function updateGroup(
  id: number,
  payload: { name?: string; description?: string | null },
): Promise<Group> {
  const { data } = await apiClient.put<Group>(`/groups/${id}`, payload);
  return data;
}

export async function deleteGroup(id: number): Promise<void> {
  await apiClient.delete(`/groups/${id}`);
}

/** 그룹원 초대 (owner/admin). role 은 admin|member (owner 직접 부여 불가). */
export async function inviteMember(
  groupId: number,
  userId: number,
  role: Exclude<GroupRole, "owner"> = "member",
): Promise<GroupMember> {
  const { data } = await apiClient.post<GroupMember>(`/groups/${groupId}/members`, {
    user_id: userId,
    role,
  });
  return data;
}

export async function removeMember(groupId: number, userId: number): Promise<void> {
  await apiClient.delete(`/groups/${groupId}/members/${userId}`);
}

/** 그룹원 역할 변경 (owner 만, admin|member). */
export async function changeMemberRole(
  groupId: number,
  userId: number,
  role: Exclude<GroupRole, "owner">,
): Promise<GroupMember> {
  const { data } = await apiClient.put<GroupMember>(
    `/groups/${groupId}/members/${userId}/role`,
    { role },
  );
  return data;
}

/** 소유권 이전 (owner 만, 대상은 활성 멤버). */
export async function transferOwnership(groupId: number, userId: number): Promise<Group> {
  const { data } = await apiClient.post<Group>(`/groups/${groupId}/transfer-ownership`, {
    user_id: userId,
  });
  return data;
}
